### ISSUE TITLE
Authentication failure with url.git@github.com:.insteadOf configuration

### ISSUE THREAD

**Issue by Deewiant** [RustSec/rustsec-crate#62]
This should be fairly simple to reproduce with cargo-audit, at least I hope it's not too tied to my personal setup:

```sh
git config --global url.git@github.com:.insteadOf https://github.com/
git init test
cd test
cargo init
cargo build
cargo audit
```

The above sequence works up until `cargo audit`, which says:

```
    Fetching advisory database from `https://github.com/RustSec/advisory-db.git`
error: couldn't fetch advisory database: git operation failed: failed to authenticate when downloading repository
attempted ssh-agent authentication, but none of the usernames `git` succeeded
```

Without the global `url.git@github.com:.insteadOf` configuration there are no issues.

I use this kind of configuration to avoid having to deal with HTTPS username/password authentication in GitHub, using my SSH setup instead, but it seems that the authentication handling within this library can't deal with it. Cloning with `git clone https://github.com/RustSec/advisory-db.git` works just fine, as expected.

**Comment by tarcieri** [comment]
I'd suggest filing an issue against the upstream `git2` crate: https://github.com/rust-lang/git2-rs

**Comment by Deewiant** [comment]
I dug around a bit and I guess my problem is in the "only supports using ssh-agent" part here:

https://github.com/RustSec/rustsec-crate/blob/300ff398bcba62e3c06e4e37b891b9a4b7d253e5/src/repository/authentication.rs#L75-L78

If I make sure to launch `ssh-agent` and `ssh-add` my key explicitly, it works. I would've expected it to ask interactively (or "just work" in the case of an un-passphrased key) but I guess that's a feature request for `git2` then. It does have [`CredentialType.SSH_INTERACTIVE`](https://docs.rs/git2/0.8.0/git2/struct.CredentialType.html#associatedconstant.SSH_INTERACTIVE) but there's no support code for it as far as I can tell, so I don't know how it's supposed to be used.

Feel free to close this since it indeed doesn't seem like it's your problem.

**Comment by tarcieri** [comment]
Aha! That's some cargo... culted (sorry pun) code from cargo.

Do you have similar issues with cargo itself? If not, it's possible it was fixed upstream.

**Comment by Deewiant** [comment]
Ha, that's a good question! I actually meant to mention this in my original report but by the time I got to writing it I had forgotten: I've got the following setting in my `~/.cargo/config`:

```
[net]
git-fetch-with-cli = true
```

I got it from https://github.com/rust-lang/cargo/issues/1851, which points to https://github.com/rust-lang/cargo/issues/1306 which has now been closed as stale but is basically the exact "no support for interactive SSH authentication" issue which is the underlying problem here as well. Perhaps you could cargo-cult some more code from Cargo to support this, although having this crate rely on Cargo's config doesn't seem like the best idea...

**Comment by tarcieri** [comment]
I believe this was addressed by #78 and #94. Please reopen if the issue persists.
